strawberryfields/backends/bosonicbackend/bosoniccircuit.py

```diff
@@ -376,79 +433,121 @@ def measure_dyne(self, covmat, indices, shots=1):
+
+        expind = np.concatenate((2 * np.array(indices), 2 * np.array(indices) + 1))
+        vals = np.zeros((shots, 2 * len(indices)))
+        pos_weights_ind = np.where(np.array(self.weights).real > 0)[0]
+        pos_weights = np.array(self.weights)[pos_weights_ind].real
+        pos_weights = pos_weights / np.sum(pos_weights)
+
+        for i in range(shots):
+            drawn = False
+            while drawn == False:
+                peak_ind_sample = np.random.choice(pos_weights_ind, size=1, p=pos_weights)[0]
+
+                cov_meas = self.covs[peak_ind_sample, expind, :][:, expind] + covmat
+                peak_sample = np.random.multivariate_normal(
+                    self.means[peak_ind_sample, expind].real, cov_meas.real
+                )
+
+                exp_arg = np.einsum(
+                    "...j,...jk,...k",
+                    (peak_sample - self.means[:, expind]),
+                    np.linalg.inv(self.covs[:, expind, :][:, :, expind] + covmat),
+                    (peak_sample - self.means[:, expind]),
+                )
+                weighted_exp = np.array(self.weights) * np.exp(-exp_arg)
+                prob_dist_val = np.sum(weighted_exp)
+                prob_upbnd = np.sum(weighted_exp[pos_weights_ind])
+                vertical_sample = np.random.random(size=1) * prob_upbnd
+                if vertical_sample < prob_dist_val:
+                    drawn = True
+                    vals[i] = peak_sample
+                if drawn == True:
+                    break
+
+        # The next line is a hack in that it only updates conditioned on the first samples value
+        # should still work if shots = 1
+        if len(indices) < len(self.active):
+            self.post_select_generaldyne(covmat, indices, vals[0])
+
+        return vals
 
     def homodyne(self, n, shots=1, eps=0.0002):
         """Performs a homodyne measurement by calling measure dyne an giving it the
         covariance matrix of a squeezed state whose x quadrature has variance eps**2"""
-        # covmat = np.diag(np.array([eps ** 2, 1.0 / eps ** 2]))
-        # res = self.measure_dyne(covmat, [n], shots=shots)
+        covmat = self.hbar * np.diag(np.array([eps ** 2, 1.0 / eps ** 2])) / 2
+        res = self.measure_dyne(covmat, [n], shots=shots)
+        return res
+
+    def heterodyne(self, n, shots=1):
+        """Performs a homodyne measurement by calling measure dyne an giving it the
+        covariance matrix of a squeezed state whose x quadrature has variance eps**2"""
+        covmat = self.hbar * np.eye(2) / 2
+        res = self.measure_dyne(covmat, [n], shots=shots)
+        return res
+
+    def post_select_generaldyne(self, covmat, indices, vals):
+        """ Performs a generaldyne measurement but postelecting on the value vals for modes n """
+        if covmat.shape != (2 * len(indices), 2 * len(indices)):
+            raise ValueError("Covariance matrix size does not match indices provided")
+
```

Comment:
do we want to add a check to make sure covmat is a valid covariance matrix here ?

Reply:
I'll add this change when I make the other covmat change. Thanks!